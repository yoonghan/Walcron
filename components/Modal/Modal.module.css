.container {
  position: fixed;
  z-index: var(--z-index-modal);
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  background-color: rgba(100, 100, 100, 0.8);
  cursor: pointer;
}

.container > button {
  text-align: right;
  color: var(--dialog-text, #fff);
  background: transparent;
  border: none;
}

.content {
  position: fixed;
  z-index: var(--z-index-modal-content);
  animation-name: slideIn;
  animation-duration: 0.4s;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  min-width: 320px;
}
