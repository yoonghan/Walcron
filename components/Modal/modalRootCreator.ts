function create() {
  if (document.getElementById("modal-root") === null) {
    const portalRoot = document.createElement("div")
    portalRoot.setAttribute("id", "modal-root")
    document.body.appendChild(portalRoot)
  }
}

const props = { create }

export default props
