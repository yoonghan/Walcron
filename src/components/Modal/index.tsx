`use strict`

/**
  Displaying the dialog for help
  **/

import * as React from "react";
import {ESC} from "../../shared/keyboardkey";
import {DIALOG} from "../../shared/style";

interface ModalProps {
  ignoreSelfClose?: boolean;
  cancelCallback: (event?: React.MouseEvent<HTMLElement, MouseEvent>) => void;
}

class Modal extends React.PureComponent<ModalProps, {}> {
  private dialogContainerRef = React.createRef<HTMLDivElement>();
  private escRef = React.createRef<HTMLButtonElement>();
  private keyListenerEvent: (evt: KeyboardEvent) => void;
  private documentBody = document.getElementsByTagName("body")[0];

  constructor(props:ModalProps) {
    super(props);
    const _cancelCallback = props.cancelCallback;
    this.keyListenerEvent = function (evt: KeyboardEvent) {
      if ( (evt.keyCode as any) === ESC )   {
         _cancelCallback();
      }
    };
  }

  componentDidMount() {
    if(this.escRef.current !== null) {
      this.escRef.current.focus();
    }
    this.documentBody.addEventListener('keyup', this.keyListenerEvent);
  }

  componentWillUnmount() {
    this.documentBody.removeEventListener('keyup', this.keyListenerEvent);
  }

  _onClickContainer = (event: React.MouseEvent<HTMLElement>) => {
    const {ignoreSelfClose, cancelCallback} = this.props;
    if(ignoreSelfClose) {
      if(event && this.dialogContainerRef && event.target === this.dialogContainerRef.current) {
        cancelCallback(event);
      }
    }
    else {
      cancelCallback(event);
    }
  }

  render() {
    return (
      <div className={"modal"} onClick={this._onClickContainer} ref={this.dialogContainerRef}>
        <div className={"modal-content"}>
          {this.props.children}
        </div>
        <div className={"buttonContainer"} onClick={this.props.cancelCallback}>
          <button ref={this.escRef}>[ESC]</button>
        </div>
        <style jsx> {`
          button {
            color: ${DIALOG.FOREGROUND};
            background: transparent;
            border: none;
          }
          .buttonContainer {
            text-align: right;
          }
          .modal {
            position: fixed;
            z-index: 99;
            left: 0;
            top: 0;
            width: 100%;
            height: 100%;
            overflow: auto;
            background-color: rgba(100,100,100,0.8);
            cursor: pointer;
          }
          .modal-content {
            position: fixed;
            z-index: 100;
            animation-name: slideIn;
            animation-duration: 0.4s;
            top: 50%;
            left: 50%;
            transform: translate(-50%, -50%);
            min-width: 320px;
          }
        `}</style>
      </div>
    );
  }
}

export default Modal;
