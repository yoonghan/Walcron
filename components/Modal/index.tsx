/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/tabindex-no-positive */
import { createRef, useEffect } from "react"
import { createPortal } from "react-dom"
import styles from "./Modal.module.css"
import modalRootCreator from "./modalRootCreator"

interface ModalProps {
  isModal?: boolean
  onCancel: (
    event?: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>
  ) => void
  children: React.ReactNode
}

const Modal = ({ isModal = false, onCancel, children }: ModalProps) => {
  modalRootCreator.create()
  const escRef = createRef<HTMLButtonElement>()
  const dialogElem = createRef<HTMLDialogElement>()
  const documentModal = document.querySelector("#modal-root") as Element

  useEffect(() => {
    if (escRef.current !== null) {
      escRef.current.focus()
    }
  }, [escRef])

  useEffect(() => {
    if (!isModal) {
      return onCancel
    }
  }, [isModal, onCancel])

  useEffect(() => {
    if (dialogElem.current !== null && !dialogElem.current.open) {
      if (isModal) {
        dialogElem.current.show()
      } else {
        dialogElem.current.showModal()
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isModal])

  return createPortal(
    <dialog className={styles.container} ref={dialogElem}>
      <div className={styles.content}>{children}</div>
      <button ref={escRef} onClick={onCancel}>
        [ESC]
      </button>
    </dialog>,
    documentModal
  )
}

export default Modal
