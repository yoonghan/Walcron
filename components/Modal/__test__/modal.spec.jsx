import * as React from 'react';
import { shallow } from 'enzyme';
import Modal from "../index";

describe('Modal', () => {
  it('renders normally', () => {
    const wrapper = shallow(
      <Modal
        ignoreSelfClose={true}
        cancelCallback={()=>{}}
      />
    );
    expect(wrapper.debug()).toMatchSnapshot();
  }),
  it('renders non-ignored normally', () => {
    const wrapper = shallow(
      <Modal
        ignoreSelfClose={false}
        cancelCallback={()=>{}}
      />
    );
    expect(wrapper.debug()).toMatchSnapshot();
  })
});
