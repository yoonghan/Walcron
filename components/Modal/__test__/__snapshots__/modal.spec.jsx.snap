// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders non-ignored normally 1`] = `
"<div onClick={[Function (anonymous)]} className=\\"jsx-1217570507 modal\\">
  <div className=\\"jsx-1217570507 modal-content\\" />
  <div onClick={[Function: cancelCallback]} className=\\"jsx-1217570507 buttonContainer\\">
    <button className=\\"jsx-1217570507\\">
      [ESC]
    </button>
  </div>
  <JSXStyle id=\\"3525693946\\" dynamic={{...}}>
    button.__jsx-style-dynamic-selector{color:#FFF;background:transparent;border:none;}.buttonContainer.__jsx-style-dynamic-selector{text-align:right;}.modal.__jsx-style-dynamic-selector{position:fixed;z-index:99;left:0;top:0;width:100%;height:100%;overflow:auto;background-color:rgba(100,100,100,0.8);cursor:pointer;}.modal-content.__jsx-style-dynamic-selector{position:fixed;z-index:100;-webkit-animation-name:slideIn-__jsx-style-dynamic-selector;animation-name:slideIn-__jsx-style-dynamic-selector;-webkit-animation-duration:0.4s;animation-duration:0.4s;top:50%;left:50%;-webkit-transform:translate(-50%,-50%);-ms-transform:translate(-50%,-50%);transform:translate(-50%,-50%);}
  </JSXStyle>
</div>"
`;

exports[`Modal renders normally 1`] = `
"<div onClick={[Function (anonymous)]} className=\\"jsx-1217570507 modal\\">
  <div className=\\"jsx-1217570507 modal-content\\" />
  <div onClick={[Function: cancelCallback]} className=\\"jsx-1217570507 buttonContainer\\">
    <button className=\\"jsx-1217570507\\">
      [ESC]
    </button>
  </div>
  <JSXStyle id=\\"3525693946\\" dynamic={{...}}>
    button.__jsx-style-dynamic-selector{color:#FFF;background:transparent;border:none;}.buttonContainer.__jsx-style-dynamic-selector{text-align:right;}.modal.__jsx-style-dynamic-selector{position:fixed;z-index:99;left:0;top:0;width:100%;height:100%;overflow:auto;background-color:rgba(100,100,100,0.8);cursor:pointer;}.modal-content.__jsx-style-dynamic-selector{position:fixed;z-index:100;-webkit-animation-name:slideIn-__jsx-style-dynamic-selector;animation-name:slideIn-__jsx-style-dynamic-selector;-webkit-animation-duration:0.4s;animation-duration:0.4s;top:50%;left:50%;-webkit-transform:translate(-50%,-50%);-ms-transform:translate(-50%,-50%);transform:translate(-50%,-50%);}
  </JSXStyle>
</div>"
`;
